### PR TITLE
Feature/shader set tess geometry

### PR DESCRIPTION
**Pull request by norru**
Testing currently in progress, still WIP. It would be nice to have a mantainer to look at the changes so far.

PR checklist:
- [x] cargo build and cargo test on Linux/OpenGL pass
- [ ] `make reftests` succeeds (?)
- [x] tested examples with the following backends: gl
- [x] `rustfmt` run on changed code (validation only)

### CODE REVIEW

File: .project [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
```
Comment (kvark) [inline .project]: was this accidentally added? you may want to add ".project" to ".gitignore" instead

File: src/core/src/lib.rs [hunk 2]
```diff
@@ -117,6 +117,8 @@ pub enum ShaderSet<R: Resources> {
     Geometry(VertexShader<R>, GeometryShader<R>, PixelShader<R>),
     /// Tessellated TODO: Tessellated, TessellatedGeometry, TransformFeedback
     Tessellated(VertexShader<R>, HullShader<R>, DomainShader<R>, PixelShader<R>),
+    /// Tessellated TODO: Tessellated, TessellatedGeometry, TransformFeedback
```
Comment (kvark) [inline src/core/src/lib.rs]: let's remove the TODO from here and above comment

File: src/render/src/factory.rs [hunk 3]
```diff
@@ -234,6 +234,37 @@ pub trait FactoryExt<R: Resources>: Factory<R> {
         Ok(ShaderSet::Tessellated(vs, hs, ds, ps))
     }
 
+    /// Creates a `ShaderSet` from the supplied vertex, hull, domain, geometry and pixel
+    /// shader source code. Mainly used for testing.
+    fn create_shader_set_tessellation_with_geometry(&mut self, vs_code: &[u8], hs_code: &[u8], ds_code: &[u8], gs_code: &[u8], ps_code: &[u8])
+                         -> Result<ShaderSet<R>, ProgramError> {
+        let vs = match self.create_shader_vertex(vs_code) {
```
Comment (kvark) [inline src/render/src/factory.rs]: let's be a tiny bit more modern and write this as `let vs = self.create_shader_vertex(vs_code).map_err(ProgramError::Vertex)?;`

File: src/core/src/lib.rs (hunk 2 above)
Comment (norru) [reply src/core/src/lib.rs]: Done

File: examples/terrain_tessellated/main.rs [hunk 4]
```diff
@@ -127,12 +131,23 @@ impl<R: gfx::Resources> gfx_app::Application<R> for App<R> {
...
 
         let (vbuf, slice) = factory.create_vertex_buffer_with_slice(&vertex_data, &index_data[..]);
-        let set = factory.create_shader_set_tessellation(
-            &vs.select(backend).unwrap(),
-            &hs.select(backend).unwrap(),
-            &ds.select(backend).unwrap(),
-            &ps.select(backend).unwrap()
-        ).unwrap();
+
+		let set = if let Ok(gs) = gs.select(backend) {
```
Comment (kvark) [inline examples/terrain_tessellated/main.rs]: indentation seems off?

File: examples/terrain_tessellated/shader/terrain.glslg [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	V v;
+	v.pos = v_In[i].pos;
+	v.color = v_In[i].color;
+	return v;
+}
+
+void emit_vertex(V v) {
+	v_Color =  v.color;
+	gl_Position = u_Proj * u_View * u_Model * vec4(v.pos, 1.0);
+    gl_ClipDistance[0] = 1.0;
```
Comment (kvark) [inline examples/terrain_tessellated/shader/terrain.glslg]: nit: indentation

Comment (norru) [reply examples/terrain_tessellated/shader/terrain.glslg]: Argh, ffs. I don't have autoformat for glsl. Will fix manually.

File: examples/terrain_tessellated/main.rs (hunk 4 above)
Comment (norru) [reply examples/terrain_tessellated/main.rs]: Gonna run cargo fmt in a minute. Tidying up now.

File: src/render/src/factory.rs (hunk 3 above)
Comment (norru) [reply src/render/src/factory.rs]: Done for all the `create_shader_*` convenience calls.

File: examples/terrain_tessellated/main.rs (hunk 4 above)
Comment (norru) [reply examples/terrain_tessellated/main.rs]: `cargo fmt` kinda does a mess - indented manually instead.

File: examples/terrain_tessellated/shader/terrain.glslg (hunk 5 above)
Comment (norru) [reply examples/terrain_tessellated/shader/terrain.glslg]: Fixed indentation.

File: .project (hunk 1 above)
Comment (norru) [reply .project]: Removed `.project` and added `.gitignore` line

**Comment by norru** [comment]
Hi @kvark, any chance to get a new `pre-ll` release on `crates.io` any time soon?

**Comment by kvark** [comment]
This is unfortunately an API change, so we'd be talking about 0.18 here, and with that we need to go through the API and double check all the other potential fixes.

**Comment by norru** [comment]
Ping - would you be able to double check about the 0.18 release?

**Comment by kvark** [comment]
Let's merge in https://github.com/gfx-rs/gfx/pull/2197, then update dependencies, and we'll be good to go. It's easier to coordinate this on gitter, would you be able to join the channel (https://gitter.im/gfx-rs/gfx)?

**Pull request merged**